### PR TITLE
Issue 8: Vault and Strategy Holdings

### PR DESCRIPTION
**Pull request by miguel567**
Issue: https://github.com/iearn-finance/yearn-api/issues/8

- Added ./abi folder as a volume to docker-compose.yml file to ease development.
- calculated Vault holdings and Strategy holdings and Pools Balance (virtualPrice * Totalsupply)
- Storing price_usd from coingecko for vault's want tokens.
- Created a new table called holdings.

I'm following the approach from dudesahn as defined here: hackmd.io/@dudesahn/BkxKfTzqw
TODO: We believe we are missing to calculate the pool balance of yWBTC token.
This is the pre-requirement for calculating TVL (my next step).

### CODE REVIEW

**Comment by miguel567** [comment]
I'm breaking APY/SAVE. need to fix it first.

**Comment by miguel567** [comment]
fixed the APY save endpoint which was broken before and removed ethereum-blocks-by-date.js file as this is not needed.

File: services/vaults/apy/save/handler.js [hunk 1]
```diff
@@ -276,7 +276,7 @@ const readVault = async (vault) => {
   return data;
 };
 
-module.exports.handler = async () => {
+const handler = async () => {
```
Comment (ghost) [inline services/vaults/apy/save/handler.js:279]: why change this? was it broken?

File: README.md [hunk 2]
```diff
@@ -32,37 +32,6 @@ git clone https://github.com/yearn-integrations/yearn-api.git
...
-- Run the command `docker-compose up` to test API endpoints locally
-- You can reach the API under localhost:3000
-  - use localhost:3000/local or localhost:3000/dev, depending on your environment
-  - If you want to change the local port, change the "ports" entry under "serverless" in the docker-compose.yml
-- The dirs config, services and utils are mounted into the running container, so you code changes will become available instantly on the running instance
-  - If you want to change that, remove the "volumes" entries under "serverless" in the docker-compose.yml
-
-## Stages
-
-- Currently four stages are available
```
Comment (ghost) [inline README.md]: what is intention behind removing this README section?

File: services/vaults/holdings/save/handler.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    price_id: price_id
+  } = vault;
+  console.log(`Reading vault ${vault.name}`);
+  if (!abi || !address) {
+    console.log(`Vault ABI not found: ${name}`);
+    return null;
+  }
+  try{
+    const holdings= await getHoldings(vault);
+    const priceFeed = await axios.get('https://api.coingecko.com/api/v3/coins/' + vault.price_id);
```
Comment (ghost) [inline services/vaults/holdings/save/handler.js]: can we get price on-chain from uniswap instead perhaps? this way we don't have to maintain the API

File: services/vaults/holdings/save/vaults.js [hunk 4]
```diff
@@ -0,0 +1,293 @@
+'use strict';
+
+const config = require('./config.js');
+
+
+module.exports = [
+  {
```
Comment (ghost) [inline services/vaults/holdings/save/vaults.js:7]: Can we get vaults from vaults endpoint instead? The vaults endpoint pulls from the vault registry

File: services/vaults/save/handler.js [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
+
 require('dotenv').config();
 const _ = require('lodash');
 const dynamodb = require('../../../utils/dynamoDb');
 const fetch = require('node-fetch');
 const Web3 = require('web3');
 const yRegistryAbi = require('../../../abi/yRegistry.json');
 const delay = require('delay');
-
+const db = dynamodb.doc;
```
Comment (ghost) [inline services/vaults/save/handler.js:12]: nice fix, i spotted this as well

File: services/vaults/apy/save/handler.js (hunk 1 above)
Comment (miguel567) [reply services/vaults/apy/save/handler.js:279]: At the end of the file I have set:
module.exports = { handler, getVirtualPrice };

The reason is that I'm reusing the same getVirtualPrice function in : https://github.com/miguel567/yearn-api/blob/91c1696f5f5fa309126883fd17a51ee964a1448b/services/vaults/holdings/save/getHoldings.js#L94

so I had to modify the exports module to include both functions.

File: README.md (hunk 2 above)
Comment (miguel567) [reply README.md]: I totally believe this is a mistake. Shoud I fix it and push again?

File: services/vaults/holdings/save/handler.js (hunk 3 above)
Comment (miguel567) [reply services/vaults/holdings/save/handler.js]: OK, I need to investigate how to perform that. Any guidance?

File: services/vaults/holdings/save/vaults.js (hunk 4 above)
Comment (miguel567) [reply services/vaults/holdings/save/vaults.js:7]: ah yes, I saw you improved that part, I think I can do this easily.

File: services/vaults/holdings/save/handler.js (hunk 3 above)
Comment (miguel567) [reply services/vaults/holdings/save/handler.js]: Can we do this on an improvement ticket? Unless it is a blocker, I could make the other improvements easily, deliver the TVL endpoint and then focus on changing the price feed.

File: services/vaults/holdings/save/vaults.js (hunk 4 above)
Comment (miguel567) [reply services/vaults/holdings/save/vaults.js:7]: Icannot use vaults from DB as the ABI is not stored. I could use:
`
const fetchAbi = async (address) => {
  const url = `https://api.etherscan.io/api?module=contract&action=getabi&address=${address}&apikey=${etherscanApiKey}`;
  const resp = await fetch(url).then((res) => res.json());
  const abi = JSON.parse(resp.result);
  await delay(delayTime);
  return abi;
};
`
from services/vaults/holdings/save/handler.js, but we would rely on etherscan API.

File: services/vaults/holdings/save/getHoldings.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    symbol,
+    erc20address,
+    decimals
+  } = vault;
+  const strategyAddress = await getVaultsStrategy(vault);
+  const strategyContract = new web3.eth.Contract(strategyMinABI, strategyAddress);
+  const vaultContract = new web3.eth.Contract(vault.vaultContractABI, vault.vaultContractAddress);
+  const vaultHoldings = await vaultContract.methods.balance().call() / Math.pow(1,vault.decimals);
+  const strategyHoldings = await strategyContract.methods.balanceOf().call() / Math.pow(1,vault.decimals);
+    /* console.log(holdings); */
```
Comment (graham-u) [inline services/vaults/holdings/save/getHoldings.js]: Can we remove commented out console.log?

File: services/vaults/holdings/save/getHoldings.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        holdings = {
+          'strategyAddress': strategyAddress,
+          'vaultHoldings': vaultHoldings,
+          'strategyHoldings': strategyHoldings    
+        } 
+      }
+
+
+    return holdings;  
+}
```
Comment (graham-u) [inline services/vaults/holdings/save/getHoldings.js]: Need an empty line at end of file.

File: services/vaults/holdings/save/handler.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  await db
+    .put(params)
+    .promise()
+    .catch((err) => console.log("err", err));
+  console.log(`Saved ${data.name}`);
+};
+
+
+
+
```
Comment (graham-u) [inline services/vaults/holdings/save/handler.js]: Personal opinion but I think 4 blank lines here is a lot, better to have 1 or 2.  Same above as well.

**Review by graham-u — COMMENTED** [review]
There are quite a few instances of commented out code that I think we should remove unless it is there for a good reason.

File: README.md (hunk 2 above)
Comment (miguel567) [reply README.md]: solved

File: services/vaults/holdings/save/handler.js (hunk 3 above)
Comment (miguel567) [reply services/vaults/holdings/save/handler.js]: the "earn" products: yWBTC, yCRV and ybCRV are using getVirtualPrice = on-chain price.
Rest of tokens are using coingecko. I would merge as it is and create a new issue to move all pricing infro to on-chain using chainlink, for example. issue: https://github.com/iearn-finance/yearn-api/issues/34

File: services/vaults/holdings/save/vaults.js (hunk 4 above)
Comment (miguel567) [reply services/vaults/holdings/save/vaults.js:7]: As discussed in discord this cannot be done for now. We cannot fetch the ABI on-chain. Also, the services/vaults/holdings/save/vaults.js file is defining where the ABI is coming from. Also I cannot remove the vault.js file as, for instance the aLink and Link strategies have 6 decimals and there is no on-chain "decimals" method for these strategies for me to fetch the correct decimals. So for now we need to rely on the code in the files and not on-chain data :S

File: services/vaults/holdings/save/getHoldings.js (hunk 6 above)
Comment (miguel567) [reply services/vaults/holdings/save/getHoldings.js]: Updated in the new push all the commented out code.

File: services/vaults/holdings/save/getHoldings.js (hunk 7 above)
Comment (miguel567) [reply services/vaults/holdings/save/getHoldings.js]: done.

File: services/vaults/holdings/save/handler.js (hunk 8 above)
Comment (miguel567) [reply services/vaults/holdings/save/handler.js]: removed

**Comment by miguel567** [comment]
I've reviewed your feedback @graham-u  and @x48-crypto  from the previous push, please check it above.

Also I've added to this push:
- YFI staked holdings calculations
- veCRV holdings calculations.
- Added the compound Vault holdings calculations (I will add it to APY endpoint in this issue: https://github.com/iearn-finance/yearn-api/issues/32)


FYI: I'm following Dudesahn's documented holdings calculation guide: https://hackmd.io/@dudesahn/BkxKfTzqw

File: services/vaults/holdings/save/handler.js (hunk 3 above)
Comment (dudesahn) [reply services/vaults/holdings/save/handler.js]: We should be good, we can just use Andre's keep3r price oracles, which already have all of the pairs we need. I updated with the relevant info on issue #34

**Review by ghost — APPROVED** [review]

**Pull request merged**